File: src/SemanticTypeForgers/forger.tests.go
package forger

import (
	"strings"
	"testing"
)

func TestForger(t *testing.T) {
	behavior := NewAtomicBehavior[any]("test")

	if behavior.GetPrimitiveType(5) != 5 {
		t.Errorf("Test failed")
	}

	if behavior.ConvertToPrimitive(" 123 ") != 123 {
		t.Errorf("Test failed")
	}

	if behavior.ConvertToPrimitive("true") != true {
		t.Errorf("Test failed")
	}

	if behavior.ConvertToPrimitive("{}") != 0 {
		t.Errorf("Test failed")
	}

	val := behavior.ProcessValue(map[string]any{
		"productPrice":  "10.5",
		"deliveryPrice": 2.5,
	})

	if val != 13.0 {
		t.Errorf("Test failed: expected 13.0 got %v", val)
	}

	emailBehavior := NewAtomicBehavior[any]("PersonEmail")
	if emailBehavior.Validate("test@example.com") != true {
		t.Errorf("Test failed: valid email")
	}
	if emailBehavior.Validate("invalid-email") != false {
		t.Errorf("Test failed: invalid email")
	}

	defer func() {
		if r := recover(); r == nil {
			t.Errorf("Test failed: should have thrown on invalid forge")
		} else {
			if s, ok := r.(string); ok && !strings.Contains(s, "Validation failed") {
				t.Errorf("Test failed: wrong panic message: %v", s)
			}
		}
	}()
	emailBehavior.Forge("invalid-email")
}

File: src/SemanticTypeForgers/forger.tests.hs
module Main where

import Forger
import qualified Data.Map as Map
import Control.Monad (unless)

main :: IO ()
main = do
    unless (getPrimitiveType 5 == (5 :: Int)) $ error "Test failed"

    unless (convertToPrimitive (PString " 123 ") == PInt 123) $ error "Test failed"
    unless (convertToPrimitive (PString "true") == PBool True) $ error "Test failed"
    unless (convertToPrimitive (PString "{}") == PInt 0) $ error "Test failed"

    let m = Map.fromList [("productPrice", PString "10.5"), ("deliveryPrice", PDouble 2.5)]
    unless (processValueMap m == PDouble 13.0) $ error "Test failed"

    unless (validate "PersonEmail" (PString "test@example.com") == True) $ error "Test failed"
    unless (validate "PersonEmail" (PString "invalid-email") == False) $ error "Test failed"

    putStrLn "Haskell tests passed"
